File: flutter_audio/android/src/main/kotlin/com/remix/flutter_audio/model/Song.kt
package com.remix.flutter_audio.model

import android.content.ContentUris
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.provider.MediaStore

data class Song(
  val id: Long,
  val displayName: String,
  val title: String,
  val album: String,
  val albumId: Long,
  val artist: String,
  val artistId: Long,
  private var _duration: Long,
  val data: String,
  val size: Long,
  val year: String,
  private var _genre: String?,
  val track: String?,
  val dateModified: Long) : java.io.Serializable{
  
  val duration: Long
    get() {
      if (_duration <= 0 && id > 0 && data.isNotEmpty()) {
        val metadataRetriever = MediaMetadataRetriever()
        try {
          metadataRetriever.setDataSource(data)
          _duration =
            metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)!!
              .toLong()
        } catch (ignore: Exception) {
        } finally {
          metadataRetriever.release()
        }
      }
      return _duration
    }
  
  val genre: String
    get() {
      if (_genre.isNullOrEmpty() && id > 0 && data.isNotEmpty()) {
        val metadataRetriever = MediaMetadataRetriever()
        try {
          metadataRetriever.setDataSource(data)
          _genre = metadataRetriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_GENRE)!!
        } catch (ignore: Exception) {
        } finally {
          metadataRetriever.release()
        }
      }
      return _genre ?: ""
    }
  
  val contentUri: Uri
    get() = ContentUris.withAppendedId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, id)
  
  val artUri: Uri
    get() = ContentUris.withAppendedId(Uri.parse("content://media/external/audio/albumart/"), albumId)

  fun toMap(): Map<String, Any?> {
    val map = HashMap<String, Any?>()
    map["id"] = id
    map["displayName"] = displayName
    map["title"] = title
    map["album"] = album
    map["albumId"] = albumId
    map["artist"] = artist
    map["artistId"] = artistId
    map["duration"] = _duration
    map["data"] = data
    map["size"] = size
    map["year"] = year
    map["genre"] = _genre
    map["track"] = track
    map["dateModified"] = dateModified

    return map
  }

  companion object {
    @JvmStatic
    val EMPTY_SONG = Song(-1, "", "", "", -1, "", -1, -1, "", -1, "", "", "", -1)
  }
}

File: flutter_audio/lib/models/song.dart
class Song {
  final num id;
  final String displayName;
  final String title;
  final String album;
  final num albumId;
  final String artist;
  final num artistId;
  final num duration;
  final String data;
  final num size;
  final String year;
  final String? genre;
  final String? track;
  final num dateModified;

//<editor-fold desc="Data Methods">
  const Song({
    required this.id,
    required this.displayName,
    required this.title,
    required this.album,
    required this.albumId,
    required this.artist,
    required this.artistId,
    required this.duration,
    required this.data,
    required this.size,
    required this.year,
    this.genre,
    this.track,
    required this.dateModified,
  });

  String get uri => "content://media/external/audio/media/$id";
  String get artUri => "content://media/external/audio/media/$id/albumart";

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      (other is Song &&
          runtimeType == other.runtimeType &&
          id == other.id);

  @override
  int get hashCode =>
      id.hashCode;

  @override
  String toString() {
    return 'Song{' +
        ' id: $id,' +
        ' displayName: $displayName,' +
        ' title: $title,' +
        ' album: $album,' +
        ' albumId: $albumId,' +
        ' artist: $artist,' +
        ' artistId: $artistId,' +
        ' duration: $duration,' +
        ' data: $data,' +
        ' size: $size,' +
        ' year: $year,' +
        ' genre: $genre,' +
        ' track: $track,' +
        ' dateModified: $dateModified,' +
        '}';
  }

  Song copyWith({
    num? id,
    String? displayName,
    String? title,
    String? album,
    num? albumId,
    String? artist,
    num? artistId,
    num? duration,
    String? data,
    num? size,
    String? year,
    String? genre,
    String? track,
    num? dateModified,
  }) {
    return Song(
      id: id ?? this.id,
      displayName: displayName ?? this.displayName,
      title: title ?? this.title,
      album: album ?? this.album,
      albumId: albumId ?? this.albumId,
      artist: artist ?? this.artist,
      artistId: artistId ?? this.artistId,
      duration: duration ?? this.duration,
      data: data ?? this.data,
      size: size ?? this.size,
      year: year ?? this.year,
      genre: genre ?? this.genre,
      track: track ?? this.track,
      dateModified: dateModified ?? this.dateModified,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'displayName': displayName,
      'title': title,
      'album': album,
      'albumId': albumId,
      'artist': artist,
      'artistId': artistId,
      'duration': duration,
      'data': data,
      'size': size,
      'year': year,
      'genre': genre,
      'track': track,
      'dateModified': dateModified,
    };
  }

  factory Song.fromMap(Map<dynamic, dynamic> map) {
    return Song(
      id: map['id'] as num,
      displayName: map['displayName'] as String,
      title: map['title'] as String,
      album: map['album'] as String,
      albumId: map['albumId'] as num,
      artist: map['artist'] as String,
      artistId: map['artistId'] as num,
      duration: map['duration'] as num,
      data: map['data'] as String,
      size: map['size'] as num,
      year: map['year'] as String,
      genre: map['genre'] as String?,
      track: map['track'] as String?,
      dateModified: map['dateModified'] as num,
    );
  }

//</editor-fold>


}
